### PR TITLE
Link plugins against their dependencies - rebased to master

### PR DESCRIPTION
**Pull request by bluelightning32**
Android does not properly support the RTLD_GLOBAL flag to dlopen. So .cmxs
plugins need to be linked against all other plugins they get symbols from.

This is https://github.com/coq/coq/pull/6020 rebased against the master branch.

### CODE REVIEW

**Comment by Zimmi48** [comment]
FTR you could have rebased without creating a new PR, just force-push and edit the base branch (edit button next to the PR title).

**Comment by ejgallego** [comment]
I think we've understood the problem, thanks. Please a bit of patient while we review this change, it will take some time for us as this is a bit untypical workflow.

**Comment by Zimmi48** [comment]
> FTR you could have rebased without creating a new PR, just force-push and edit the base branch (edit button next to the PR title).

Oh I see that you wanted to preserve your branch (just read your last comment in the other PR). Well, indeed there is no way to change the branch of a PR so there was no way of avoiding closing the other one and reopening this one.

**Comment by ejgallego** [comment]
> Oh I see that you wanted to preserve your branch (just read your last comment in the other PR). Well, indeed there is no way to change the branch of a PR so there was no way of avoiding closing the other one and reopening this one.

What I do in this is I just duplicate my branch locally.

**Comment by maximedenes** [comment]
Who could review this?

**Comment by Zimmi48** [comment]
@letouzey Do you feel competent to review this? Who else could do it?

File: Makefile.build
Comment (SkySkimmer) [inline Makefile.build]: This seems like it should be moved to a script file, eg tools/make-cmxs-d.sh

Comment (SkySkimmer) [reply Makefile.build]: Or maybe it's an extension of ocamllibdep, I don't know enough about that thing to tell.

Comment (bluelightning32) [reply Makefile.build]: The best place to make the change would be ocamldep, but you're using the version from ocaml, and I don't see any chance of getting those changes accepted there.

I don't want to put it in ocamllibdep. The output from ocamllibdep is already exposed as variables. Parsing a Makefile inside ocaml seems even more ugly than in sed.

I can move it to a sh script file as you suggested.

**Comment by ejgallego** [comment]
> @letouzey Do you feel competent to review this? Who else could do it?

I suggest we ask for help in https://discuss.ocaml.org/ , this looks like a tricky issue for us.

**Comment by Zimmi48** [comment]
@ejgallego I agree. I had been similarly thinking of calling for help on Coq-Club but Discuss OCaml is probably a good idea. Can you send a message there?

**Comment by Zimmi48** [comment]
@ejgallego Did you share the request for reviewers?

**Comment by ejgallego** [comment]
No, I didn't do it yet, this is pretty far back in my todo queue [and not a 2 minutes job] so please feel free to take over. [If I do, I will update the PR]

**Comment by herbelin** [comment]
I'm not an expert in Makefiles as others are but the patch looks well-designed to me, and not intrusive wrt non-Android platform (I suspect that even the added `LINK_PLUGINS` in `coqMakefile.in` is redundant but not a problem on non-Android platform, right?).

I'd be in favor of integrating the patch as this does not exclude that other improvements are done later on. Being supported on Android is important I think.

**Comment by ejgallego** [comment]
@herbelin do you understand the OCaml-side of the patch? IMHO we should understand it better before merging, asking on discuss cannot hurt.

**Comment by herbelin** [comment]
> do you understand the OCaml-side of the patch? IMHO we should understand it better before merging, asking on discuss cannot hurt.

You mean why we have to add the `cmx` explicitly? I'm not a specialist of `dlopen`, but isn't this, according to @bluelightning32, a problem of Android rather than of OCaml?

Of course, if OCaml can fix this upstream, that would be wonderful, and they have to find a fix somehow, and we have to report about it to them, but I guess that we shall not wait for OCaml 4.06++ before supporting Android, no?

**Comment by ejgallego** [comment]
I guess the problem is nobody around know what the patch does so it is hard to say yes/no .

File: Makefile.build
Comment (SkySkimmer) [reply Makefile.build]: >I can move it to a sh script file as you suggested.

Please do.

File: Makefile.install
Comment (SkySkimmer) [inline Makefile.install]: Configure should probably check for `patchelf` on android.

**Comment by SkySkimmer** [comment]
From `man ocamlopt`
```
       -a     Build  a  library  (.cmxa/.a  file) with the object files (.cmx/.o files) given on the
              command line, instead of linking them into an executable file. The name of the library
              must be set with the -o option.

              If  -cclib or -ccopt  options are passed on the command line, these options are stored
              in the resulting .cmxa library.  Then, linking with this  library  automatically  adds
              back  the  -cclib and -ccopt options as if they had been provided on the command line,
              unless the -noautolink option is given. Additionally, a substring $CAMLORIGIN inside a
               -ccopt  options  will be replaced by the full path to the .cma library, excluding the
              filename.
```
This seems related somehow but I don't know enough about the difference between cmxa and cmxs to tell if we could switch.

**Comment by bluelightning32** [comment]
Re cmxa,

That's for statically linking a library. That won't help because, 1) coq plugins are dynamic by definition, so we can't link them all statically into coqc. 2) If the plugins are statically linked against their dependent plugins, multiple copies of those dependent plugins would get loaded.

**Comment by ejgallego** [comment]
> coq plugins are dynamic by definition, so we can't link them all statically into coqc.

Note that this is supported and I do that often. Just in case that's useful for you. A different story is that linking many plugins will, without a doubt, create havoc and they may not interact properly.

**Comment by Zimmi48** [comment]
The comments by @SkySkimmer haven't been addressed, and we didn't find anyone yet ready to approve this, so let's postpone to 8.7.2. Sorry about that.

**Comment by Zimmi48** [comment]
@gasche Would you feel competent (and have time available) to review this? Or do you know someone who would?

**Comment by gasche** [comment]
I don't know enough about dynamic linkers to give a fully confident assessment of the proposed change, but I will (1) give some opinion anyway, (2) ping @xavierleroy who does know enough about dynamic linkers and (3) re-explain the proposed change below.

Some opinion (1): I think that it is reasonable to merge the patch; it is carefully written to not affect non-Android system, it is an improvement for Android systems. On the other hand, I am not convinced that it is a long-term fix to Android because the Android people seem to be on a war against dynamic libraries -- see this [more recent blog post](https://android-developers.googleblog.com/2016/06/android-changes-for-ndk-developers.html) that seems to indicate that even DT_NEEDED entries (which are used by the patch instead of RTLD_GLOBAL) may become unsupported in the future. My understanding is that the Android developers have decided, long-term, to break dynamic loading on their platform except for their own SDK; this PR by @bluelightning32 may thus only be the beginning of a long series of arcane changes to react to various Android breakage in the future.

Now for summary of the patch (3) for others to assess.

The patch deals with the case where `main.ml` dynamically links `a.cmxs` and then `b.cmxs`, where the OCaml module B (in `b.cmxs`) depends on `A` (in `a.cmxs`) -- this happens when Coq plugins depend on each other. Previously, `a.cmxs` and `b.cmxs` would be compiled independently and not know of each other, and then RTLD_GLOBAL would do the magic at Dynlink time. Now RTLD_LOCAL is assumed, and instead `b.cmxs` is compiled with the `-cclib a.cmxs` option, which makes things work again.

More precisely, the compilation and linking Scheme of Coq before this patch would be as follows in this scenario:

```shell
ocamlopt -verbose -shared a.ml -o a.cmxs
# produces a.s (and a.cmxs.startup.s, elided here) and runs the following commands:
as -o a.o a.s
gcc -shared -o a.cmxs -L.../lib/ocaml a.o 

ocamlopt -verbose -shared b.ml -o b.cmxs
# produces b.s (and b.cmxs.statup.s, elided here) and runs the following commands:
as -o b.o b.s
gcc -shared -o b.cmxs -L.../lib/ocaml b.o 

ocamlopt -verbose dynlink.cmxa main.ml -o main
# produces main.s (and camlstartup.s, elided here) and runs:
as -o main.o main.s
gcc -o main -L.../lib/ocaml main.o .../lib/ocaml/{dynlink,stdlib,libasmrun}.a -lm  -ldl
```

With this standard compilation scheme, b.cmxs is produced without
knowledge of a.cmxs (`objdump -p b.cmxs` does not mention a.cmxs).  We
rely on RTLD_GLOBAL for symbol resolution to happen as desired at
dynlink time, though the global namespace -- this works because
`main.ml` dynlinks `a.cmxs` and only then `b.cmxs`.

The proposed patch changes the second step (the production of b.cmxs) to:

```shell
ocamlopt -verbose -shared b.ml -cclib a.cmxs -o b.cmxs # NEW: -cclib .a.cmxs
# produces b.s (and b.cmxs.statup.s, elided here) and runs the following commands:
as -o b.o b.s
gcc -shared -o b.cmxs -L.../lib/ocaml b.o a.cmxs # NEW: a.cmxs
```

With this proposed compilation scheme, b.cmxs has a.cmxs in its NEEDED
table (`objdump -p b.cmxs` shows it), and it will correctly
re-use/share the previously-dynlinked a.cmxs.

I understand how the patch works as described, but I still find it
a bit surprising: the documentation of RTLD_LOCAL says that the symbol
of a.cmxs will not be made available to subsequently loaded shared
objects, but apparently this effect is overriden when b.cmxs links
against a.cmxs. I couldn't find enough information online to
understand whether this is intentional, or a shared implementation
detail of our linkers. The archive
[src.tar.gz](https://github.com/coq/coq/files/1413455/src.tar.gz)
provided by the report allowed me to check that it works fine on my
machine when using LD_LOCAL -- the OCaml compiler still uses LD_GLOBAL
so it cannot be used to reliably test this.

To conclude:

- in #6020 @bluelightning32 provided useful tarballs of [C files](https://github.com/coq/coq/files/1413455/src.tar.gz) and [OCaml files](https://github.com/coq/coq/files/1417136/ocaml-example.tar.gz) with compilation scripts to test all that (I added the `-verbose` flag to `ocamlopt` invocations, and used `RTLD_NOW` instead of `RTLD_LAZY` in the C files to be closed to OCaml's `caml_dlopen`)

- The complex part of the patch is a sed script that reads `ocamldep` output to recover which `.cmxs` files should be passed as extra `-cclib` flags when compiling a plugin. I have not reviewed it in details but it looks reasonable.

**Comment by ejgallego** [comment]
Thansk @gasche , so it seems that another alternative for supporting Coq on Android is just to link all plugins statically? This looks feasible to me, and it is easy to do in `master`, but I wonder what others and @bluelightning32 thinks about it?

File: Makefile.build
Comment (Zimmi48) [reply Makefile.build]: I concur. Assuming that the discussion converges toward merging this PR, having this part in a separate shell script would make it more readable and more maintainable.

**Comment by gasche** [comment]
Yes, I think that using static linking would also work for plugins distributed upstream. If you want to have third-party plugins, with one plugin depending on another, you need this work again -- but maybe that situation is less common.

If you have the ability to maintain a static build of Coq as a clearly-documented option of the build system, that could be an easy way to let people build Coq on non-standard systems (like Android here, but also systems where dynamic linking is not available for OCaml).

**Comment by ejgallego** [comment]
The main use case to maintain static linking is easy debugging and also that depending on LTAC is so common that it makes sense to link it statically for most of the alternative top-level consumers. I think that LTAC should be the main dependency of 3rd party plugins so indeed Android could choose to link it statically [as for example SerAPI does].

**Comment by Zimmi48** [comment]
A decision should be taken on this PR, either merge or close. @gasche said it was probably safe to merge but not the most durable solution. I'm personally a bit worried about merging this without any tests to ensure that we don't start breaking it and without anyone understanding enough about it to maintain it correctly. So I vote to close.

@bluelightning32 What do you think of the static linking solution?

**Comment by ejgallego** [comment]
@gasche comments wrt the future of Android dynamic linking are worrying, @bluelightning32 , what do you think? Is dynamic linking going to really become frowned upon on Android?

If so, I'd vote as Théo to close and recommend static linking instead; that should work for most cases.

**Comment by ejgallego** [comment]
A downside of static linking is that some plugins may "conflict", but indeed, IMHO that's a bug in the plugin and should be fixed. That being said, I regularly use static builds and with the common set of plugins that works OK.

**Comment by bluelightning32** [comment]
I'm okay with you closing this PR instead of merging it. I haven't seen many other people comment on this thread. So it looks like Coq on Android is not an important feature at this point. Given that, supporting Android is probably not worth the complexity it adds to the build scripts.

Even if this PR is closed, it would still come up in searches, and the work could be resumed if Coq on Android becomes more important in the future.

**Comment by ejgallego** [comment]
@bluelightning32 I am not sure if I understand what you mean but indeed even if Coq in Android is not a primary platform now I think developers would be happy to support it. I think the discussion is more on the lines of "How do we support Coq on Android?" than in "Do we support Coq on Android".

So we have two choices:
- merge this PR,
- adapt Coq on Android to use static linking [that's easy], then close this PR.

The first choice would indeed be good, however it seems that it may not be the most sensible strategy due to Android platform policies. WDYT?

**Comment by bluelightning32** [comment]
How would static linking work if you want to install a 3rd party Coq plugin through opam? Would the 3rd party plugin relink Coq? How would you uninstall a plugin?

**Comment by ejgallego** [comment]
> How would static linking work if you want to install a 3rd party Coq plugin through opam? Would the 3rd party plugin relink Coq? How would you uninstall a plugin?

Indeed it would provide a new coq binary with the plugin linked `coqtop.plugin`; I do something similar in my toplevels [that do require static linking but for different reasons] however in a hackish way. Once we have Dune as a build option it will be more principled.

I think we all much prefer dynamic linking of course, but if that's gonna stop working in the next Android release, then I am not sure it is worth the effort. But if you think that's the way to go, let's indeed go ahead.

**Comment by bluelightning32** [comment]
> I think we all much prefer dynamic linking of course, but if that's gonna stop working in the next Android release, then I am not sure it is worth the effort. But if you think that's the way to go, let's indeed go ahead.

I think the blog post that gasche linked to refers to linking against system libraries that are not part of the public API. I'm not aware of any reason that dynamically loading against shared libraries within the same application would stop working.

**Comment by gasche** [comment]
@bluelightning32, I think there may be a minor communication issue here: asked about it by Coq developers, I tried to provide [an assessment](https://github.com/coq/coq/pull/6033#issuecomment-363728589) of your PR (highlighting the fact that I'm far from an expert on this area), and to explain my reasoning at each step so that one can easily point at a specific thing I wrote and said "actually that's not right, ..." and provide an explanation.

I was hoping that you (or other people more knowledgeable that would chime in on the PR, but that didn't happen) would comment on whether the assessment sounds right, or what is wrong in it -- I wouldn't be surprised if some comments are misguided. But you haven't replied to any of the things I wrote so far, and instead you are now writing that "So it looks like Coq on Android is not an important feature at this point" -- a subjective comment with no actionable item.

I think it would be useful for the Coq developers (not me) to have an idea of what you think of the discussion of the PR that was made. Do you agree with the points raised, or do you think that there are incorrect? More precisely:

1. I commented that dynamic linking third-party libraries that are not part of the Android NDK may stop working at some point (to a non-expert it looks like this is what the [blog post](https://android-developers.googleblog.com/2016/06/android-changes-for-ndk-developers.html) I linked above is saying). Do I correctly understand from your [last comment](https://github.com/coq/coq/pull/6033#issuecomment-384070934) that this is not the case? Will it remain possible to dynlink third-party libraries provided by plugin authors with the Coq executable? What is the difference between this use-case and what the blog post describes?

2. I indicated that I found the approach surprising, as I don't understand (again, as a non-expert) how RTLD_LOCAL works when a plugin `.cmxs` depends on another plugin's `.cmxs`, as the documentation I could find didn't indicate explicitly that this is possible. Would you mind explaining (to non-experts) why and how this work, and whether this is a standard/specified feature or something that may change in the future?

3. There was a suggestion to rewrite the 27-lines long sed script as a separate bash script, to make the changes to Makefile.build less invasive. I think it would be much easier for maintainers to think something like "I don't know how/why this works, but changing a non-currently-working build in a non-invasive way is fine as long as it doesn't increase my maintenance burden" if the diff to the Makefile was smaller (by factoring this shell script out). Have you considered refining your PR with such a change, which shouldn't be much work?

I understand that doing contributions that require specialized skills and having them reviewed very slowly by people that know much less about the area can be frustrating. But I also have the impression that the difficulty you perceive in getting changes accepted (for example when you "don't see any chance of getting those changes accepted [in ocamldep]", which has not been my experience) may also be caused by the fact that you haven't been particularly effective, yet, in explaining what your work is doing to people that are not as familiar with the tools and abstraction level you are working at.

**Comment by ejgallego** [comment]
Thanks @gasche for the help and nice summary; @bluelightning32 , please bear with us, your contribution is very much welcome and appreciated, it is just that indeed the area this PR touches seems complex and outside of our expertise so we are very slow as we'd like to document what's going on so for example, when we merge Dune support, we know what has to be done for Android to work.

**Comment by bluelightning32** [comment]
@gasche 

> I commented that dynamic linking third-party libraries that are not part of the Android NDK may stop working at some point (to a non-expert it looks like this is what the blog post I linked above is saying). Do I correctly understand from your last comment that this is not the case? Will it remain possible to dynlink third-party libraries provided by plugin authors with the Coq executable? What is the difference between this use-case and what the blog post describes?

Okay... I'm guessing you keep asking me because my commit has a google.com email address. I work for Google on something not related to Coq, Ocaml, or Android. I use Coq as a hobby outside work. The way my employment contract is written, there are fewer hoops for me jump through if I send PRs with a google.com address.

I asked someone who works on the Android team. He clarified that where that blog post says, "applications are no longer able to load non-NDK platform libraries," it means non-(NDK) platform libraries. It does not mean non-(NDK platform) libraries. He said that many apps have executables that link against .so libraries from the same app. So that should continue to work.

Also, the changes in that blog post already took affect in Android N. I tested my Coq patch against Android O.

> I indicated that I found the approach surprising, as I don't understand (again, as a non-expert) how RTLD_LOCAL works when a plugin .cmxs depends on another plugin's .cmxs, as the documentation I could find didn't indicate explicitly that this is possible. Would you mind explaining (to non-experts) why and how this work, and whether this is a standard/specified feature or something that may change in the future?

RTLD_LOCAL prevents the symbols from getting added to the global namespace. However, the symbols can still be accessed if the library is loaded again.

The [POSIX 1003.1](http://pubs.opengroup.org/onlinepubs/009695399/functions/dlopen.html) standard says, "Only a single copy of an object file is brought into the address space, even if dlopen() is invoked multiple times in reference to the file."

> There was a suggestion to rewrite the 27-lines long sed script as a separate bash script

It sounds like this PR is getting close to being merged. I'll work moving the sed script out this week.

> I understand that doing contributions that require specialized skills and having them reviewed very slowly by people that know much less about the area can be frustrating.

I understand there is very little intersection between the people who have the skills used in this PR. For example, I don't know Ocaml, and I muddled my way though that part of debugging the build process.

I already accomplished my primary goal with this PR, which was to share the changes necessary to run Coq on Android. I haven't been frustrated, but my interest in this PR did wane during the lengthy review process.

**Comment by ejgallego** [comment]
Thanks to all, my understanding of the situation is much better now.

Something to try before merging is if the Dune PR ( #6857 ) + [cross compiling](https://github.com/ocaml/dune/blob/master/doc/advanced-topics.rst) can produce a working Coq.

If so, that could save some tedious Makefile scripting.

**Comment by ejgallego** [comment]
Note that the Dune port requires 4.06.0 [on purpose] but the Ocaml Android seems to be stuck at 4.04.0

I'll happily port `coq_dune` to 4.04.0 if there is not a newer Android compiler so we can try.

**Comment by ejgallego** [comment]
Ok, I've updated #6857 so the Dune method can be tried, need to switch to other stuff but I'll keep an eye on this. [In particular I'd need to know what Android architecture I should try, and any other build instructions]

**Comment by Zimmi48** [comment]
@ejgallego I'm not sure I can follow what you are saying about the interaction of this PR and Dune. Can you try to explain better? @bluelightning32 may have even less context and may not be able to follow either.

**Comment by ejgallego** [comment]
Sorry @Zimmi48 , Dune does support cross compilation and understands how to properly generate correct Android executables; see the above links. I was able to produce a Coq build for Android just by setting the proper cross-compiling target, but I lack the means to test such binary.

If that works, it'd mean that this PR wouldn't be needed once the OPAM packages use Dune for build.

Of course, the amount of insight this PR has provides is amazing.

**Comment by gasche** [comment]
@bluelightning32 thanks for the clarifications. I had no idea that you worked for Google, but I am glad to know that you are now even more informed of Android dynamic linking status than we could reasonably hope for.

**Comment by bluelightning32** [comment]
The .ml.d files were merged together into plugins/.mlfiles.d. My PR doesn't work when I merge it to master. Fixing it is going beyond the capabilities of sed.

I'll try to make an ocaml script that parses plugins/.mlfiles.d and plugins/.mllibfiles.d and to build a new file with the cmxs dependencies.

**Comment by ejgallego** [comment]
> I'll try to make an ocaml script that parses plugins/.mlfiles.d and plugins/.mllibfiles.d and to build a new file with the cmxs dependencies.

An OCaml developer just said above [IIUC] that this is not the case; why don't we try to modify indeed ocaml instead of doing a hack on the files it generates?

**Comment by bluelightning32** [comment]
> An OCaml developer just said above [IIUC] that it is not the case that ocamldep is closed to changes; why don't we try to modify indeed ocaml instead of doing a hack on the files it generates?

In that case, @ejgallego and @gasche, please review my design before I start muddling through ocaml code and trying to build a PR for ocamldep.

## Background
The .ml files are compiled into .cmo and .cmx files. An .ml file often depends on an interface exported by another compiled file (cmx, or .cmi). Ocamldep parses the .ml files to determine which order they should be built in so that those interface are available at the right time. That order is serialized into plugins/.mlfiles.d as makefile-style rules.

The .mlpack files describe how multiple .cmx files are grouped together into a module (.cmxa/.cmxs). ocamllibdep translates all the .mlpack file from a newline separated lists of modules into plugins/.mllibfiles.d, which is a makefile. There is one variable that ends with _MLPACK_DEPENDENCIES for each of the .mlpack files. Each _MLPACK_DEPENDENCIES variable has a list of the .cmx files (with the .cmx extension stripped) that should be linked together into a .cmxs file. ocamllibdep also includes rules to build the _plugin.cmx files. The _plugin.cmx rules use makefile-style foreach loops to iterate through all the items in the corresponding _MLPACK_DEPENDENCIES variable, then appends a .cmx suffix to them.

The _plugin.cmx files are linked into _plugin.cmxs files by a rule in Makefile.build.

## Proposed changes to Makefile.build
The rule in Makefile.build that creates the .cmxs needs to know all the other .cmxs files it depends on, so that it can link them with the -cclib option. It happens that if any of the .ml files the .cmxs was built from depended on a _plugin.cmx file, then the corresponding _plugin.cmxs needs to be linked in.

So Makefile.build first needs to iterate all of the .ml files the .cmxs was built with. Fortunately that is accessible from the _MLPACK_DEPENDENCIES variable. Next it needs to search the dependencies of those .ml files for dependencies that end with _plugin.cmx. Unfortunately the _plugin.cmx dependencies were stored as makefile rules, not makefile variables. So that is not accessible from Makefile.build.

Previously I was parsing the dependencies from the .ml.d files using a sed script, then omitting a new makefile with the dependencies as variables.

## Proposed changes to ocamldep
I would add a new argument to ocamldep that would do this:
  -var_suffix <suffix> In addition to saving the dependencies as makefile rules, create a makefile variable for each target, with a list of its dependencies.

**Comment by ejgallego** [comment]
Thanks @bluelightning32 , I will review ASAP.

**Review by gares — APPROVED** [review]
The change to CoqMakefile.in is OK.

The entire patch is, to my eyes, over complicated for what it achieves (Android support). I'll articulate more in another message.

**Comment by gares** [comment]
With my brand new hat of supposed-maintainer of the makefiles, my honest impression is that the patch is too complex for the benefits it provides.

I don't know exactly what @bluelightning32  is trying to do, but I imagine he is building an android app with Coq. If so I think that the option of statically linking all plugins should be preferred: you build a monolitic coqtop with all plugins linked in (the standard ones + a selection of your choice). Would that be an option @bluelightning32 ? I guess Android user would just download your bundle and be happy. Sure, this patch would make the app support add-ons, but is it really that important?

The Makefile even used to have support for statically linking all plugins in the plugin directory, I don't recall if it is still the case. Even if the capability had to be restored I'm sure it would be simpler: you just (re)link coqtop adding to it all plugin's `.cmxa` files.

**Comment by gares** [comment]
You are right.

I did a quick read at the proposed design. It seems an improvement wrt the current patch, but what I propose still seems an order of magnitude simpler (assuming the limitations of the proposed approach are acceptable). Also patching `ocamldep` adds quite some "bureaucracy" to the process (eg we could use the new feature in what, Coq 8.10?).

The only thing that one could salvage is the (automatic) calculation of the correct order of plugin's .cmxa files to be linked in (that, IIRC just contain the single, packed, `_plugin.cmx` file). I was thinking one could do that by hand, since it does not change much. One could the wait the eventual fix to ocamldep to have this extra feature.

I think only @bluelightning32 can tell if the simpler approach I propose is viable for his objective.
If so this patch becomes 3 lines in the main makefiles (to link all coq plugins in, say, bin/coqtop.static) and eventually the make rule can honor a variable EXTRA_CMXA_TO_LINK to make the rule extensible to plugins external to the Coq source tree.

**Comment by gares** [comment]
Oh, I forgot one last detail. Some external plugins do (probably all of them should) ship a `META` file (for their .cmxa archive). Linking these cmxa is even simpler since you can just use `-package unicoq,mtac2,...` to link the "augmented" coqtop.

**Comment by bluelightning32** [comment]
I was not trying to build a full android app. I was trying to get coq to compile in termux (an android app). The termux maintainer never accepted my opam patches. So there does not seem to be enough need for coq on android support.

Also the gnurootdebian app is being maintained again. That provides a more linux like environment that does not require changes to the coq build file. So that seems like a better way to go.

**Comment by ejgallego** [comment]
Thanks a lot @bluelightning32 for your interest in Coq, I'm sorry we weren't able to process this PR better; I hope that if it is useful in the future we can still start from here.

Note that I had some success cross-compiling Coq using Dune, so that could be an interesting way to pursue.

**Pull request closed without merging**
